src/cart.c

```diff
@@ -51,10 +51,17 @@ static int detect_format(struct cartridge *self, FILE *f)
 static int parse_ines(struct cartridge *self, FILE *f)
 {
     unsigned char header[16];
+    size_t items_expected, items_read = 0, item_size;
 
-    fread(header, sizeof header[0], sizeof header, f);
-    if (feof(f)) return CART_ERR_EOF;
-    if (ferror(f)) return CART_ERR_IO;
+    item_size = sizeof header[0];
+    items_expected = sizeof header;
+    while (items_read < items_expected) {
+	items_read += fread(&header[item_size * items_read], item_size,
+			    items_expected - items_read, f);
+
+	if (feof(f)) return CART_ERR_EOF;
+	if (ferror(f)) return CART_ERR_IO;
+    }
```

Comment:
I'm not clear this loop is needed and I'm curious if you've run into cases where that's necessary. reading `fread`'s man page or cppreference (https://en.cppreference.com/w/c/io/fread) it seems that you'd only get a return value < `count` if either EOF or an error occurred. I suppose it could happen if `FILE *f` ends up being a pipe or other buffered stream?

Reply:
I think there's a case where it returns less than the expected count, but not zero, due to a signal interruption or asynchronous cancel.

So retrying makes sense.  But on review, your code is honestly a cleaner model, both in understanding and fail fast behavior.  If there are changes to the file in-between calls to fread, my code returns inconsistent data.

Thanks